collector/cpu_linux.go

```diff
@@ -0,0 +1,160 @@
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build !nocpu
+
+package collector
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/procfs"
+)
+
+const (
+	cpuCollectorNamespace = "cpu"
+)
+
+type cpuCollector struct {
+	cpu                *prometheus.Desc
+	cpuFreq            *prometheus.Desc
+	cpuFreqMin         *prometheus.Desc
+	cpuFreqMax         *prometheus.Desc
+	cpuCoreThrottle    *prometheus.Desc
+	cpuPackageThrottle *prometheus.Desc
+}
+
+func init() {
+	Factories["cpu"] = NewCPUCollector
+}
+
+// NewCPUCollector returns a new Collector exposing kernel/system statistics.
+func NewCPUCollector() (Collector, error) {
+	return &cpuCollector{
+		cpu: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, "", cpuCollectorNamespace),
+			"Seconds the cpus spent in each mode.",
+			[]string{"cpu", "mode"}, nil,
+		),
+		cpuFreq: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, cpuCollectorNamespace, "frequency_hertz"),
+			"Current cpu thread frequency in hertz.",
+			[]string{"cpu"}, nil,
+		),
+		cpuFreqMin: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, cpuCollectorNamespace, "frequency_min_hertz"),
+			"Minimum cpu thread frequency in hertz.",
+			[]string{"cpu"}, nil,
+		),
+		cpuFreqMax: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, cpuCollectorNamespace, "frequency_max_hertz"),
+			"Maximum cpu thread frequency in hertz.",
+			[]string{"cpu"}, nil,
+		),
+		cpuCoreThrottle: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, cpuCollectorNamespace, "core_throttles_total"),
+			"Number of times this cpu core has been throttled.",
+			[]string{"cpu"}, nil,
+		),
+		cpuPackageThrottle: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, cpuCollectorNamespace, "package_throttles_total"),
+			"Number of times this cpu package has been throttled.",
+			[]string{"cpu"}, nil,
+		),
+	}, nil
+}
+
+// Update implements Collector and exposes cpu related metrics from /proc/stat and /sys/.../cpu/.
+func (c *cpuCollector) Update(ch chan<- prometheus.Metric) error {
+	if err := c.updateStat(ch); err != nil {
+		return err
+	}
+	if err := c.updateCPUfreq(ch); err != nil {
+		return err
+	}
+	return nil
+}
+
+// updateCPUfreq reads /sys/bus/cpu/devices/cpu* and expose cpu frequency statistics.
+func (c *cpuCollector) updateCPUfreq(ch chan<- prometheus.Metric) error {
+	cpus, err := filepath.Glob(sysFilePath("bus/cpu/devices/cpu[0-9]*"))
+	if err != nil {
+		return err
+	}
+
+	for _, cpu := range cpus {
+		_, cpuname := filepath.Split(cpu)
+
+		if value, err := readUintFromFile(filepath.Join(cpu, "cpufreq/scaling_cur_freq")); err != nil {
+			return err
+		} else {
```

Comment:
Sorry for that nitpick, but can you drop the } else { block? That's more idiomatic. Like:

```
if value, err := readUintFromFile(filepath.Join(cpu, "cpufreq/scaling_cur_freq")); err != nil {
￼    return err
}
￼ch <- prometheus.MustNewConstMetric(c.cpuFreq, prometheus.GaugeValue, float64(value), cpuname)
```